File: CSharpCodeBase/entities/ui/uigadgetcontroller.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class uigadgetcontroller :entity{
 var UIGadgetController= Class({__includes=Entity, init=function(self)            
       Entity.init(self)   ;
       this.gameObject = luanet.UnityEngine.GameObject.Find("UIGadgetsPanel");
       this.gadgetController = this.gameObject:GetComponent("UIGadgetController");
       this.gadgetController:SetupLuaController(self);
       this.gadgetController:SetCallback(this.Callback)     ;
       this.inputService = GameController.inputService;
       this.enabled = true;
       this.topGadget = {}
       this.botGadget = {}
       this.leftGadget = {}
       this.rightGadget = {}
       this.activeGadget = {}
       this.activePanel = -1;
       this.gadgetController:Hide();
       this.gadgetController:SetText("");
 }})
 public void OnGadgetButtonClick(){
   if(#this.activeGadget > 0  &&  GameController.inventory:GetItemCount(this.activeGadget[1]) > 0  ){ 
     var characterEntity = GameController.player.playerController:GetCurrentSlot();
     var path = GameController.database:Get("items", this.activeGadget[1] .."/require");
     var Gadget = require(path);
     var gadget = Gadget(characterEntity, this.activeGadget[1]);
     gadget:Apply();
   }else{
     this.activeGadget = {}
     this.activePanel = -1;
   }
 }
 public void OnInventaryChanged(){
   if(#this.activeGadget > 0  &&  GameController.inventory:GetItemCount(this.activeGadget[1]) > 0  ){ 
     this.gadgetController:SetText(GameController.inventory:GetItemCount(this.activeGadget[1]));
   }else{
     this.gadgetController:SetText("");
   }
 }
 public void Update(){
   Entity.Update(self);
   if(this.inputService:RightButtonWasPressed()  ){ 
     this.gadgetController:Hide();
   }
   
   if(this.inputService:TopButtonWasPressed()  ){ 
     self:OnGadgetButtonClick();
   }
   
   if(this.inputService:DPadRightIsPressed()  ){ 
     self:DPadClick(2);
   }else{if this.inputService:DPadLeftIsPressed()  ){ 
     self:DPadClick(1);
   }else{if this.inputService:DPadUpIsPressed()  ){ 
     self:DPadClick(0);
   }else{if this.inputService:DPadDawnIsPressed()  ){ 
     self:DPadClick(3);
   }else{if this.inputService:GetKey(KeyCode.H)  ){ 
     if(not this.gadgetController.active   ){ 
       this.gadgetController:Show();
     }
   }else{if this.gadgetController.active  ){ 
     this.gadgetController:Hide();
   }
 }
 public void DPadClick(count){
    this.activePanel = count;
    
    if(not this.gadgetController.active   ){ 
       this.gadgetController:Show();
    }
    this.gadgetController:ShowPanel(this.activePanel);
 }
 public void GetGadgets(){
   var tbl = {}
   for(k,v in pairs(GameController.inventory.grids.inventory:GetFilteredItems("gadget")) ){ 
     var subTbl = {}
     table.insert(subTbl, v.name);
     table.insert(subTbl, v.count);
     table.insert(subTbl, GameController.database:Get("items", v.name .."/middle_icon"));
     table.insert(tbl, subTbl);
   }
   return tbl;
 }
 public void OnHide(){
   this.buttonClickTime = 0;
   this.inputService:PopFrame();
 }
 public void OnShow(){
   this.inputService:PushFrame("gadgets_menu");
 }
 public void SetLeftGadget(name, count){
   this.leftGadget = {}
   table.insert(this.leftGadget, name);
   table.insert(this.leftGadget, count);
   table.insert(this.leftGadget, GameController.database:Get("items", name .."/middle_icon"));
 }
 public void SetRightGadget(name, count){
   this.rightGadget = {}
   table.insert(this.rightGadget, name);
   table.insert(this.rightGadget, count);
   table.insert(this.rightGadget, GameController.database:Get("items", name .."/middle_icon"));
 }
 public void SetTopGadget(name, count){
   this.topGadget = {}
   table.insert(this.topGadget, name);
   table.insert(this.topGadget, count);
   table.insert(this.topGadget, GameController.database:Get("items", name .."/middle_icon"));
 }
 public void SetBotGadget(name, count){
   this.botGadget = {}
   table.insert(this.botGadget, name);
   table.insert(this.botGadget, count);
   table.insert(this.botGadget, GameController.database:Get("items", name .."/middle_icon"));
 }
 public void SetActiveGadget(name, count){
   this.activeGadget = {}
   table.insert(this.activeGadget, name);
   table.insert(this.activeGadget, count);
   table.insert(this.activeGadget, GameController.database:Get("items", name .."/middle_icon"));
   this.gadgetController:SetText(count);
 }
 public void GetRightGadget(){
   return this.rightGadget;
 }
 public void GetLeftGadget(){
   return this.leftGadget;
 }
 public void GetTopGadget(){
   return this.topGadget;
 }
 public void GetBotGadget(){
   return this.botGadget;
 }
 public void GetActivePanel(){
   return this.activePanel;
 }
 }}

File: CSharpCodeBase/entities/ui/uigadgetcontroller.lua
local Class = require 'utils.hump.class'
local Entity = require 'unity.entity'
local RecipeList = require 'components.inventory.recipelist'

local UIGadgetController= Class({__includes=Entity, init=function(self)            
      Entity.init(self)   
      self.gameObject = luanet.UnityEngine.GameObject.Find("UIGadgetsPanel")
      self.gadgetController = self.gameObject:GetComponent("UIGadgetController")
      self.gadgetController:SetupLuaController(self)
      self.gadgetController:SetCallback(self.Callback)     
      self.inputService = GameController.inputService
      self.enabled = true
      self.topGadget = {}
      self.botGadget = {}
      self.leftGadget = {}
      self.rightGadget = {}
      self.activeGadget = {}
      self.activePanel = -1
      self.gadgetController:Hide()
      self.gadgetController:SetText('')
end})

function UIGadgetController:OnGadgetButtonClick()
  if #self.activeGadget > 0 and GameController.inventory:GetItemCount(self.activeGadget[1]) > 0 then
    local characterEntity = GameController.player.playerController:GetCurrentSlot()
    local path = GameController.database:Get('items', self.activeGadget[1] ..'/require')
    local Gadget = require(path)
    local gadget = Gadget(characterEntity, self.activeGadget[1])
    gadget:Apply()
  else
    self.activeGadget = {}
    self.activePanel = -1
  end
end

function UIGadgetController:OnInventaryChanged()
  if #self.activeGadget > 0 and GameController.inventory:GetItemCount(self.activeGadget[1]) > 0 then
    self.gadgetController:SetText(GameController.inventory:GetItemCount(self.activeGadget[1]))
  else
    self.gadgetController:SetText('')
  end
end

function UIGadgetController:Update()
  Entity.Update(self)
  if self.inputService:RightButtonWasPressed() then
    self.gadgetController:Hide()
  end
  
  if self.inputService:TopButtonWasPressed() then
    self:OnGadgetButtonClick()
  end
  
  if self.inputService:DPadRightIsPressed() then
    self:DPadClick(2)
  elseif self.inputService:DPadLeftIsPressed() then
    self:DPadClick(1)
  elseif self.inputService:DPadUpIsPressed() then
    self:DPadClick(0)
  elseif self.inputService:DPadDawnIsPressed() then
    self:DPadClick(3)
  elseif self.inputService:GetKey(KeyCode.H) then
    if not self.gadgetController.active  then
      self.gadgetController:Show()
    end
  elseif self.gadgetController.active then
    self.gadgetController:Hide()
  end
end

function UIGadgetController:DPadClick(count)
   self.activePanel = count
   
   if not self.gadgetController.active  then
      self.gadgetController:Show()
   end
   self.gadgetController:ShowPanel(self.activePanel)
end

function UIGadgetController:GetGadgets()
  local tbl = {}
  for k,v in pairs(GameController.inventory.grids.inventory:GetFilteredItems('gadget')) do 
    local subTbl = {}
    table.insert(subTbl, v.name)
    table.insert(subTbl, v.count)
    table.insert(subTbl, GameController.database:Get('items', v.name ..'/middle_icon'))
    table.insert(tbl, subTbl)
  end
  return tbl
end

function UIGadgetController:OnHide()
  self.buttonClickTime = 0
  self.inputService:PopFrame()
end

function UIGadgetController:OnShow()
  self.inputService:PushFrame('gadgets_menu')
end

function UIGadgetController:SetLeftGadget(name, count)
  self.leftGadget = {}
  table.insert(self.leftGadget, name)
  table.insert(self.leftGadget, count)
  table.insert(self.leftGadget, GameController.database:Get('items', name ..'/middle_icon'))
end

function UIGadgetController:SetRightGadget(name, count)
  self.rightGadget = {}
  table.insert(self.rightGadget, name)
  table.insert(self.rightGadget, count)
  table.insert(self.rightGadget, GameController.database:Get('items', name ..'/middle_icon'))
end

function UIGadgetController:SetTopGadget(name, count)
  self.topGadget = {}
  table.insert(self.topGadget, name)
  table.insert(self.topGadget, count)
  table.insert(self.topGadget, GameController.database:Get('items', name ..'/middle_icon'))
end

function UIGadgetController:SetBotGadget(name, count)
  self.botGadget = {}
  table.insert(self.botGadget, name)
  table.insert(self.botGadget, count)
  table.insert(self.botGadget, GameController.database:Get('items', name ..'/middle_icon'))
end

function UIGadgetController:SetActiveGadget(name, count)
  self.activeGadget = {}
  table.insert(self.activeGadget, name)
  table.insert(self.activeGadget, count)
  table.insert(self.activeGadget, GameController.database:Get('items', name ..'/middle_icon'))
  self.gadgetController:SetText(count)
end

function UIGadgetController:GetRightGadget()
  return self.rightGadget
end

function UIGadgetController:GetLeftGadget()
  return self.leftGadget
end

function UIGadgetController:GetTopGadget()
  return self.topGadget
end

function UIGadgetController:GetBotGadget()
  return self.botGadget
end

function UIGadgetController:GetActivePanel()
  return self.activePanel
end


return UIGadgetController
